### PR TITLE
[cuda] Enable block splitting for cuda caching memory allocator

### PR DESCRIPTION
**Pull request by qiao-bo**
Related issue = #3634

This enables the reuse of buffers with larger size.

### CODE REVIEW

**Review by strongoier — CHANGES_REQUESTED** [review]
Looks great! Only some tiny issues :-)

File: taichi/backends/cuda/cuda_caching_allocator.cpp [hunk 1]
```diff
@@ -10,25 +10,28 @@ CudaCachingAllocator::CudaCachingAllocator(Device *device) : device_(device) {
...
-  if (find_block(params.size)) {
-    auto blk = mem_blocks_.find(params.size);
-    ret = blk->second;
-    mem_blocks_.erase(blk);
+  auto size_aligned = taichi::iroundup(params.size, taichi_page_size);
+  auto it_blk = mem_blocks_.lower_bound(size_aligned);
+
+  if (it_blk != mem_blocks_.end()) {
+    size_t remaining_sz = it_blk->first - size_aligned;
+    if ((remaining_sz > 0) && (remaining_sz % taichi_page_size == 0)) {
```
[the remark this answers is not in the document]
Comment (qiao-bo) [reply taichi/backends/cuda/cuda_caching_allocator.cpp:18]: right! iroundup guarantees its always aligned, thx!

File: taichi/backends/cuda/cuda_caching_allocator.cpp [hunk 2]
```diff
@@ -10,25 +10,31 @@ CudaCachingAllocator::CudaCachingAllocator(Device *device) : device_(device) {
...
-    ret = blk->second;
-    mem_blocks_.erase(blk);
+  auto size_aligned = taichi::iroundup(params.size, taichi_page_size);
+  auto it_blk = mem_blocks_.lower_bound(size_aligned);
+
+  if (it_blk != mem_blocks_.end()) {
+    size_t remaining_sz = it_blk->first - size_aligned;
+    if (remaining_sz > 0) {
+      TI_ASSERT(remaining_sz % taichi_page_size == 0);
+      // only split if the remaining sz is page aligned
```
Comment (strongoier) [inline taichi/backends/cuda/cuda_caching_allocator.cpp:20]:
```suggestion
```

**Review by strongoier — APPROVED** [review]
LGTM!

**Pull request merged**